### PR TITLE
test(transports): multiple addresses

### LINKED ISSUE #2194: test(transport): tcp,quic should allow multiple local addresses
we should extend `basic_test` with additional test case that asserts that transport can be started with multiple addresses. 
this should work for `tcp` and `quic` and for all other transports if it works.

do this after https://github.com/vacp2p/nim-libp2p/pull/2193 is merged

---
test cases:
- default test case: where multiple addresses are provided; dial from clients and ensure that dialed connection works
example: `test_quic.nim::should allow multiple local addresse`
- cancel dialing: same as above, but here we dial (with timeout so it gets cancelled) and do not drive accept so connection is not created (dial is cancelled). in this case we need to ensure that all already established connections will still work and that connections that we attempted to dial will not work (future will be canceled). 
- cancel accept: similar case to above, here we want to ensure that allready established connections will still work and that no new connections will be made because accept fut was cancelled. example: `basic_test.nim::handle accept cancellation`
- stopping transport while accepting: - should stop transport, close all connections; future that was blocked on `accept` should throw - transport stopped
- please consider and suggest other edge cases that can happen when stopping/canceling transport/futures

### PR DESCRIPTION
**Pull request by rlve**
## Summary

Addresses #2194

Adds transport tests covering cancellation and shutdown edge cases around `accept()` and `dial()`. 

Three things:
- A dual-stack test proving that cancelling an in-flight `dial()` on one address family leaves an already-established connection fully working.
- A dual-stack test proving that cancelling a parked `accept()` doesn't disturb an established connection and doesn't spawn a stray one.
- A basic-transport test proving that stopping a transport unblocks a pending `accept()`. QUIC resolves the parked accept to nil, the others raise `TransportClosedError`. Inconsistency raised in https://github.com/vacp2p/nim-libp2p/issues/2713

Supporting this, a small `runStreamHandler` helper in `utils.nim` opens a stream on an existing muxer and runs a handler on it, so tests can reuse an established connection to prove it still exchanges data after a cancellation.

## Affected Areas

- [x] Tests

### CODE REVIEW

**Review by gmelodie — APPROVED** [review]

File: tests/libp2p/transports/stream_tests.nim [hunk 1]
```diff
@@ -101,6 +101,107 @@ template streamTransportTest*(
...
+        await stream.readExactly(addr buffer, clientMessage.len)
+        check string.fromBytes(buffer) == clientMessage
+        await stream.write(serverMessage)
+
+    proc clientStreamHandler(stream: MuxedStream) {.async: (raises: []).} =
+      noExceptionWithStreamClose(stream):
+        await stream.write(clientMessage)
+        var buffer: array[serverMessage.len, byte]
+        await stream.readExactly(addr buffer, serverMessage.len)
+        check string.fromBytes(buffer) == serverMessage
```
Comment (vladopajic) [inline tests/libp2p/transports/stream_tests.nim]: `check (await stream.readExactlyAsStr(serverMessage.len)) ==  serverMessage`

**Review by vladopajic — APPROVED** [review]
